Generalize bots to run in a serverless framework

chrismaree opened the pull request:
This PR adjusts the bots to enable them to run as single fire, one-off executions to facilitate a "serverless" type deployment. The original execution logic has not been changed and the bots can still run in the original looping style to accommodate all original implementations.

To accommodate this the following changes were made:
1) The `shouldPoll` parameter was removed from the bot interface. This is redundant given we have the `pollingDelay` param. If `pollingDelay` is set to 0 then the bot will run as a single fire execution. If this is not set then the bot defaults to 1 minute polling. Tests were updated accordingly.
2) The terminate on `!shouldPoll` logic was refined such that the delay happens _after_ this check (previously the delay would occur before the `shouldPoll` break. 
3) in addition, a `await waitForLogger(Logger);` was added to ensure that when exiting after a single fire the bot waits until all log messages are sent (Slack, in particular, can take about a second to send).
4) `The TokenBalanceClient.js` previously would set account balances to `null` on the first request and then update them after the first call to `update`. This meant that the bot would only correctly see token balances after the _second_ query loop. This was done to keep the `getCollateralBalance`, `getSyntheticBalance` and `getEtherBalance` as synchronous calls. This pattern was good for a looping based design but in the single fire case this will not work as the bot needs to know the balances immediately, not in the next loop. As a result, this PR refactors these functions to be async by retrieving the associated balances on the initial boot. 
5) a `STARTING_BLOCK_NUMBER` environment parameter was added to the monitor bot to inform where the bot should start looking from when running a query to ignore past query events.
6) Bot startup logging message either sends a `info` or `debug` depending on the `pollingDelay`. If the polling delay is 0 (serverless mode) then the bot should not send a message to slack every time it boots. This would be very noisy, rather, it should only send a debug message.

close #1618

chrismaree commented on disputer/test/index.js:
```diff
@@ -75,7 +75,7 @@ contract("index.js", function(accounts) {
 
     let errorThrown;
     try {
-      await Poll.run(address, false, 10_000, priceFeedConfig);
```
Cleaner interface now: 0 for polling delay means imidate exit

chrismaree commented on financial-templates-lib/clients/TokenBalanceClient.js:
```diff
@@ -28,23 +28,23 @@ class TokenBalanceClient {
     this.accountMonitorList = [];
   };
 
-  getCollateralBalance = address => {
```
Previously calling these functions would return a `null` value if the client had not yet been updated. This meant that client implementers had to deal with the null case and wait until the following update round to get a balance. In this update, the balances are retrieved immediately after calling these functions to accommodate the single fire serverless execution.

nicholaspai replied:
nit: same format as above

nicholaspai replied:
This function should also be in traditional non-arrow style

chrismaree commented on financial-templates-lib/test/clients/TokenBalanceClient.js:
```diff
@@ -35,66 +35,48 @@ contract("BalanceMonitor.js", function(accounts) {
   });
 
   it("Returning token balances", async function() {
-    // Should start at empty state (null) and the address should not be resolved.
```
`null` tests are not needed anymore. balances are populated immediately after calling the `get*` functions.

chrismaree commented on financial-templates-lib/test/price-feed/CryptoWatchPriceFeed.js:
```diff
@@ -48,7 +48,7 @@ contract("CryptoWatchPriceFeed.js", function(accounts) {
     networker = new NetworkerMock();
     const dummyLogger = winston.createLogger({
       level: "info",
-      transports: []
```
This change was unrelated to this PR but was added to prevent this transport from printing logs during tests. This now conforms to the rest of our testing Winston framework.

chrismaree commented on monitors/BalanceMonitor.js:
```diff
@@ -52,40 +52,40 @@ class BalanceMonitor {
     // check if their collateral, synthetic or ether balance is below a given threshold. If it is, then
     // send a winston event. The message structure is defined with the `_createLowBalanceMrkdwn` formatter.
     for (let bot of this.botsToMonitor) {
-      if (this._ltThreshold(this.client.getCollateralBalance(bot.address), bot.collateralThreshold)) {
```
This `_ltThreshold` was previously used to deal with the `null` case if a balance had yet to be resolved. As this is now async this is no longer needed.

chrismaree commented on monitors/test/BalanceMonitor.js:
```diff
@@ -86,7 +86,6 @@ contract("BalanceMonitor.js", function(accounts) {
   it("Correctly emits messages on balance threshold", async function() {
     // Update the client.
     await tokenBalanceClient.update();
-    await tokenBalanceClient.update();
```
As token balances are returned immediately after a query, there is no longer a need to update the client two times.

nicholaspai replied:
Thank god

chrismaree commented on liquidator/test/index.js:
```diff
@@ -88,15 +88,15 @@ contract("index.js", function(accounts) {
-
-  it("Responds to incoming monitor requests while bot is alive", async function() {
+  // TODO: remove this test when we remove the black box testing URL.
+  it.skip("Responds to incoming monitor requests while bot is alive", async function() {
```
This test should be removed and for now it'll get skipped.

nicholaspai replied:
You should merge master and remove this

chrismaree commented on disputer/test/index.js:
```diff
@@ -69,25 +69,15 @@ contract("index.js", function(accounts) {
     assert.isFalse(errorThrown);
-  });
-
-  it("Sets token allowances correctly", async function() {
```
This test (and the previous one) were joined as there was a lot of overlapping logic. 

Additionally, the need to introduce the `await waitForLogger(Logger);` function call in the index loop break resulted in tests failing if sequentially tested on the main index file. After much testing, this is as a result of the `logger.end()` call in the Logger.js [here](https://github.com/UMAprotocol/protocol/blob/66cfd69e046e175e39ac58a82c10d12e56f3128d/financial-templates-lib/logger/Logger.js#L38) which acts to terminate the node stream from the Winston instance. If this line is called from a testing file the following error was generated:
```
NodeError: write after end
```

The solution is to not close the stream as sequential writes result in this error. I did find a solution to this by modifying the Logger.js implementation to check if the script is running in a testing env. if it was, then the `logger.end()` command can be skipped which solved this error. However, I chose to not implement this change and rather join the tests as this pollutes the Logger JS implementation for this bug.

nicholaspai replied:
So does this mean we cannot add additional unit tests to this script?

chrismaree replied:
right now, ye. we cant add any more to this test script. This is fine for now I think.

chrismaree commented on liquidator/test/index.js:
```diff
@@ -72,34 +72,24 @@ contract("index.js", function(accounts) {
     assert.isFalse(errorThrown);
-  });
-
-  it("Sets token allowances correctly", async function() {
```
this test joining rationale is the same as the `disruptor/test/index.js`.

nicholaspai commented on disputer/index.js:
```diff
@@ -20,21 +20,25 @@ const ExpandedERC20 = artifacts.require("ExpandedERC20");
       priceFeedConfig
-    });
+    };
+    if (pollingDelay == 0) Logger.debug(LogObject);
```
nit: should add a third "-" here to be strict

nicholaspai commented on disputer/index.js:
```diff
@@ -72,11 +76,12 @@ async function run(address, shouldPoll, pollingDelay, priceFeedConfig, disputerC
-
-      if (!shouldPoll) {
+      // If the polling delay is set to 0 then the script will terminate the bot after one full run.
+      if (pollingDelay == 0) {
```
nit: same comment as above

nicholaspai commented on financial-templates-lib/clients/TokenBalanceClient.js:
```diff
@@ -28,23 +28,23 @@ class TokenBalanceClient {
   }
 
   // Delete all data within the client.
-  clearState = async () => {
```
nit: this function format should be in traditional style (e.g. clearState() {}) since its at the class level

nicholaspai commented on financial-templates-lib/clients/TokenBalanceClient.js:
```diff
@@ -77,12 +78,17 @@ class TokenBalanceClient {
+      this.tokenBalances.syntheticBalances[address] = tokenBalancesObject.syntheticBalance;
+      this.tokenBalances.etherBalances[address] = tokenBalancesObject.etherBalance;
+
+      console.log("tokenBalancesObject", tokenBalancesObject);
```
Did you mean to leave this console.log?

nicholaspai commented on liquidator/index.js:
```diff
@@ -20,25 +20,29 @@ const ExpandedERC20 = artifacts.require("ExpandedERC20");
       monitorPort
-    });
+    };
+    if (pollingDelay == 0) Logger.debug(LogObject);
```
nit: needs a third "="

nicholaspai commented on liquidator/index.js:
```diff
@@ -85,27 +89,28 @@ async function run(address, shouldPoll, pollingDelay, priceFeedConfig, monitorPo
-      message: "Monitor server is listening",
-      portNumber: portNumber
-    });
+    // const { server, portNumber } = startServer(monitorPort);
```
This should be removed once you merge latest master

nicholaspai commented on liquidator/index.js:
```diff
@@ -85,27 +89,28 @@ async function run(address, shouldPoll, pollingDelay, priceFeedConfig, monitorPo
-
-      if (!shouldPoll) {
+      // If the polling delay is set to 0 then the script will terminate the bot after one full run.
+      if (pollingDelay == 0) {
```
nit: third "=" sign

nicholaspai commented on monitors/index.js:
```diff
@@ -24,42 +24,48 @@ const ExpandedERC20 = artifacts.require("ExpandedERC20");
       medianizerPriceFeedConfig
-    });
+    };
+    if (pollingDelay == 0) Logger.debug(LogObject);
```
nit: "="

nicholaspai commented on monitors/index.js:
```diff
@@ -174,11 +181,12 @@ async function run(
-
-      if (!shouldPoll) {
+      // If the polling delay is set to 0 then the script will terminate the bot after one full run.
+      if (pollingDelay == 0) {
```
nit: "="

nicholaspai commented on monitors/index.js:
```diff
@@ -194,7 +202,13 @@ const Poll = async function(callback) {
+
+    // Block number to search for events from. If set, acts to offset the search to ignore events in the past. If not
+    // set then default to null which indicates that the bot should start at the current block number.
+    const startingBlock = process.env.STARTING_BLOCK_NUMBER ? process.env.STARTING_BLOCK_NUMBER : null;
```
You can just write `startingBlock = process.env.STARTING_BLOCK_NUMBER` since it is `null` by default

mrice32 commented on financial-templates-lib/clients/TokenBalanceClient.js:
```diff
@@ -28,23 +28,23 @@ class TokenBalanceClient {
 
-  getCollateralBalance = address => {
-    this._registerAddress(address);
+  getCollateralBalance = async address => {
```
This seems to diverge a bit from how most of the other clients work. IIUC, most use the update method to cache on-chain state and then the `get` functions just return info from that cached data. I get why this one works a little differently -- because the addresses are not known until they're passed into the `get` methods.

From a UX perspective, this client now feels like a hybrid between updating immediately on call to `get` and updating during `update()`. Thoughts on just passing in a list of addresses on construction or in the `update()` method? Not super opinionated here -- just thought it might make sense to revisit the designs on this client now that it's being used differently.

chrismaree replied:
Ye, this is a super good point and I agree that how it was implemented was not ideal as it deviates from the other clients. It is not great for them to by async. 

To deal with this I've come up with a bit of a combination solution from your feedback. I don't want to add this to the constructor as this would require the monitor bot's `index.js` to iterate over these params and feed them in at construction time. Likewise, I don't want to feed this into the `update()` function as this will give it a different interface. 

Rather, I have added a `batchRegisterAddresses()` function that called by the constructor of the client implementor to specify addresses it wants to be populated into the client. In this implementation, the `BalanceMonitor.js` tells the client to watch these specific addresses and as such, they are ready to be queried on the first `update()` call.

In my mind this is a good solution because we keep the client implementation synchronous like the others and it does not place any burden on the root of the bots.

LMKWYT :)

mrice32 replied:
SGTM, I like this solution!

chrismaree commented:
@nicholaspai I've added all your nits.
@mrice32 I've refactored the client to keep everything synchronous. 

Let me know what you guys think.

nicholaspai commented on financial-templates-lib/clients/TokenBalanceClient.js:
```diff
@@ -48,18 +48,27 @@ class TokenBalanceClient {
   };
 
+  // Checks if an address has a resolved value(has been updated past it's initialization state).
   resolvedAddressBalance = address => {
```
Should be non-arrow style function

nicholaspai commented on financial-templates-lib/clients/TokenBalanceClient.js:
```diff
@@ -48,18 +48,27 @@ class TokenBalanceClient {
 
+  // Batch register an array of addresses. This can be used by a client implementor to register addresses on
+  // construction to enable synchronous retrieval on the first loop.
+  batchRegisterAddresses = addresses => {
```
Should be non-arrow style function

nicholaspai commented on financial-templates-lib/clients/TokenBalanceClient.js:
```diff
@@ -48,18 +48,27 @@ class TokenBalanceClient {
+
+  // Loop over all account addresses in the monitor list and for each check the balances of the
+  // respective tokens and Eth balance. Store these for synchronous retrieval.
   update = async () => {
```
Should be non-arrow style function

nicholaspai commented on financial-templates-lib/clients/TokenBalanceClient.js:
```diff
@@ -68,7 +77,8 @@ class TokenBalanceClient {
-  // async function to get the three token balances directly. Does not store balances later retrieval.
+  // Async function to get the three token balances directly. Does not store balances for later retrieval.
+  // TODO: refactor this to create an array of promises for all accounts to monitor and resolve them all at once.
   getDirectTokenBalances = async account => {
```
Should be non-arrow style function

nicholaspai commented on financial-templates-lib/clients/TokenBalanceClient.js:
```diff
@@ -77,6 +87,7 @@ class TokenBalanceClient {
   };
 
+  // Add an address to the monitored address list. Balance will only update when calling the `update()` function.
   _registerAddress = address => {
```
Should be non-arrow style function

chrismaree commented:
> Thanks for addressing all the nits. I added a few more nits for you to address then LGTM!

I agree these should not be arrow styled functions but then we need to refactor the other clients as well. Right now they _all_ use this formatting. I'm not going to implement these nits as even though this does not follow our formatting standard, it should be left for a separate PR where we refactor all the clients (`ExpiringMultiPartyClient.js`, ExpiringMultiPartyEventClient.js` as well).

I have created an issue to document the need for this refactor: #1628
